source/common/tracing/lightstep_tracer_impl.cc

```diff
@@ -15,103 +16,109 @@
 namespace Envoy {
 namespace Tracing {
 
-LightStepSpan::LightStepSpan(lightstep::Span& span, lightstep::Tracer& tracer)
-    : span_(span), tracer_(tracer) {}
+namespace {
+class LightStepLogger : Logger::Loggable<Logger::Id::tracing> {
+public:
+  void operator()(lightstep::LogLevel level, opentracing::string_view message) const {
```

Comment:
I'm not familiar with `opentracing::string_view`. Can this by passed by const reference? Same Q in a bunch of other places.

Reply:
It could, but it's usually faster to take small objects like `string_view` by value.

Also, passing by value will be the convention for `std::string_view` and is what's done in the C++17 standard library. (See for example [`operator<<`](http://en.cppreference.com/w/cpp/string/basic_string_view/operator_ltlt))

Reply:
I'm not really sure how it can be faster to pass small objects by value vs. reference (would be curious if you have doc/article I can read on that), but totally fine to leave as is if that's what C++17 will do.

Reply:
The members of small objects can get split and passed in registers, so access `std::string_view::data` can be simply read from the register. If it's passed by `const std::string_view&` then it needs to go through an extra dereference to access `std::string_view::data`. You can see it in the generated assembly.

```cpp
template <typename T>
void do_not_optimize_away(T const& val) {
  asm volatile("" : : "g"(val) : "memory");
}

void f1(std::string_view s) {
  char x = *s.data();
  char y = x + (char)1;
  do_not_optimize_away(y);
}

void f2(const std::string_view& s) {
  char x = *s.data();
  char y = x + (char)1;
  do_not_optimize_away(y);
}
```
compiled with `g++ -O3 -S` shows

f1:
```
	.globl __Z2f1St17basic_string_viewIcSt11char_traitsIcEE
__Z2f1St17basic_string_viewIcSt11char_traitsIcEE:
LFB758:
	movzbl	(%rsi), %eax
	addl	$1, %eax
	ret
```
and f2:
```
	.globl __Z2f2RKSt17basic_string_viewIcSt11char_traitsIcEE
__Z2f2RKSt17basic_string_viewIcSt11char_traitsIcEE:
LFB759:
	movq	8(%rdi), %rax
	movzbl	(%rax), %eax
	addl	$1, %eax
	ret
```

Note the extra dereferencing instructions in f2.

Reply:
Got it, that makes sense. Thanks for the explanation!